### ISSUE TITLE
Enumeration fails for COM collections that don't provide class information

### ISSUE THREAD

**Issue by dnolf** [microsoft/ClearScript#146]
Hi,

I have been having an issue in using ClearScript if we have a host object the we are exposing global to a VBScript that is managed and it expose functions/properties that return COM objects.  If the Host object is a native COM object using the DirectAccess host flag, everything is fine with the returned COM objects.  It appears the the COM objects from the .NET object get wrapped when returned to the script, which also adds some overhead to all use of those objects and it seems that it causes binder failures on COM objects that implement IEnumVariant as I see exceptions that GetEnumerator cannot be found.

The Host Object looks as follows

public class ScriptHelper
{
   private ISomeInterface _memberObject;  // COM Object
   public ScriptHelper(ISomeInterface memberObject)
   {
      _memberObject = memberObject;
   }

   public object GetMemberObject()
   {
      return _memberObject;
   }
}

It is added the script as follows:

engine.AddHostObject("ScriptHelper",HostItemFlags.GlobalMembers, scriptHelper);

In the script, we run into problems with something like this:

foreach item in GetMember().SomeList

I dug around through with the debugger and it appears that when the COM object is returned from the GetMemberObject called, it is wrapped up in a HostObject and that is what is returned to the script.  Then every with the returned object gets wrapped with Binders and such instead of just raw com calling.

I tested out if I made the ScriptHelper a native C++ Com Object and in the AddHostObject passed HostItemFlags.GlobalMembers | HostItemFlags.DirectAccess and then the com objects were no longer wrapped.

After much digging I found if I changed the following function that returns the item to the script":

private object MarshalToScriptInternal(object obj, HostItemFlags flags, HashSet<Array> marshaledArraySet)

and added the following code to check if the underlying return item was a COM Object, to return the Com object directly instead of the wrapper.

    var hostObject = obj as HostObject;
            if (hostObject != null)
            {
                if (hostObject.Target.GetType().IsCOMObject)
                {
                    return hostObject.Target;
                }
            }

This seems to have resolved my issues as now the com object itself is returned to the script similar to how you are passing the object directly if using the DirectAccess host item flag.

Do you see any problem with my approach and is this something we can include in the source.

Thanks!

Dave

**Comment by dnolf** [comment]
I forgot to mention, I tested this against version 5.6

**Comment by ClearScriptLib** [comment]
Hi @dnolf,

In what situation are you seeing `TypeLoadException`?

Thanks!

**Comment by dnolf** [comment]
The exceptions occur when the script is trying to interact with the wrapped com objects that are returned from the ScriptHelper host object.

If I revert the change I did, I see some exceptions like:

Exception thrown: 'System.ArgumentException' in System.Core.dll
Argument types do not match

when this function is called:

return CreateDynamicMetaObject(arg, Expression.Constant(arg, type));

The value for arg is a ComObject and the type is a type from a Primary interop assembly, so it complains are the Types don't match.

Then if I continue I see:

Microsoft.CSharp.RuntimeBinder.RuntimeBinderException: 'The best overloaded method match for 'Microsoft.ClearScript.Util.EnumerableHelpers.GetEnumerator(System.Collections.IEnumerable)' has some invalid arguments'

By updating to 5.6.0, the type load exceptions seems to have disappeared. 

But I guess really since the host object is returning an unmanaged com object, is there any point in having it wrapped and returned to the script instead of the raw dispatch pointer?

**Comment by ClearScriptLib** [comment]
Hi @dnolf,

>If I revert the change I did, I see some exceptions like:
>
>Exception thrown: 'System.ArgumentException' in System.Core.dll
>Argument types do not match

Exceptions are often thrown internally in complex binding scenarios, but if ClearScript is incorrectly letting any of them reach your host code, we'd like to know about it.

In our initial COM testing we encountered a number of unexpected conditions that required special casing to handle acceptably, particularly when interop assemblies were involved. Nearly every nontrivial COM object we tested had unique quirks that broke script binding in one way or another. It's quite possible that your scenario is also unique in some respect and requires a tweak within ClearScript.

Are your COM objects public? We'd love to do some testing and iron out this issue.

>But I guess really since the host object is returning an unmanaged com object, is there any point in having it wrapped and returned to the script instead of the raw dispatch pointer?

"Direct access" is a great solution with very limited applicability. It can only work with JScript/VBScript, and only if the object fully supports `IDispatch`. Modern script engines don't support COM, and countless COM APIs don't support `IDispatch`, so proxies like ClearScript's `HostItem` and `ScriptItem` are usually required.

Still, if direct access works for you, then by all means use it. Your proposed change makes it automatic, and that's probably not the right thing for ClearScript, but we'll see if we can come up with a way to enable it outside of `AddHostObject`.

But again, if you can think of any way for us to reproduce your issue, please let us know!

Thanks!

**Comment by dnolf** [comment]
I will create a full sample application that demonstrates the issue.  My current project has too many dependencies, so I will create a ATL COM C++ set of objects and a small c# project.  

One thing I noticed is you have some some host engine flags:  WindowsScriptEngineFlags

There is one in there for Marshalling decimals to the CurrencyWrapper, so maybe an additional flag could be useful so that this would be optional.

Thanks again for the quick responses and all the amazing work in this library!

**Comment by dnolf** [comment]
Finally able to reproduce it in a small sample.  I was going crazy as I couldn't reproduce it at first and had to dig to see what the difference was in my COM objects of my main application and my sample.

What it boils down to is if you have a Object that has a property that is another object and if the other object does not have a coclass, ie, not com creatable , but it has a dual interface.

The the Target object in the HostObject says __ComObject and the Type property shows something like this:  

  | Name | Value | Type
-- | -- | -- | --
▶ | Type | {Name = "IChildList" FullName = "CwiNoShowHelperLib.IChildList"} | System.Type {System.RuntimeType}

If the Object has a CoClass in the TypeLibrary, then the Type shows the Class.

In my case the object of the IChildList implements IEnumVariant, so I get an exception from about GetEnumerator.

But everything works fine if there is a coclass.

I have created a github repo with a Readme so you can reproduce the issue here:

https://github.com/dnolf/ClearScriptEnumeratorIssue

Be curious to  your findings.

Thanks!

**Comment by ClearScriptLib** [comment]
Hi @dnolf,

Thanks for providing the sample! With your help we've reproduced the issue and are testing a fix.

As you suspected, the root cause is that your implementation of the key interface (`IChildList`) doesn't provide class information.

The framework does generate an `IEnumerable` implementation, presumably because `IChildList` includes a `DISPID_NEWENUM` member, but without class information the runtime can't construct sufficient metadata for the method call binder to find it.

This is exactly the sort of unpredictable issue we encountered with nearly every real-world COM API we tested. COM's flexibility can be a problem :)

Thanks again for your help!
